### DISCUSSION TITLE
Using Sessions in WebTestCase not working

### DISCUSSION THREAD

**Discussion by DigitalTimK** [symfony/symfony#45662] [Q&A] [answered]
## Summary
Yes, I know that `session` was removed and `requestStack` shall be used. Which I did in my controller and services. But since I upgraded to `Symfony 5.4` I get a deprecation message on my PHPUnit functional tests too:

```text
Since symfony/framework-bundle 5.3: Accessing the "session" service directly 
from the container is deprecated, use dependency injection instead
```

thrown by this:
```php
 $session  = $this->client->getContainer()->get('session');
```

## Details

```php
class BookingControllerTest extends WebTestCase {

    // ...

    public function testToAddItemToCart()
    {
        $this->client  = static::createClient();
        self::ensureKernelShutdown();

        $item = (ProductFactory::random())->object();

        // make REQUEST (which adds stuff into the session)
        $this->client->request('GET', "/cart/add/{$item->getSlug()}");

        // assert RESPONSE
        $this->assertSame(302, $this->client->getResponse()->getStatusCode());
        $this->assertTrue($this->client->getResponse()->isRedirect('/cart/show'));

        // get cart from SESSION
        $session  = $this->client->getContainer()->get('session');
        $cart     = $session->get('cart');

        // assert content of cart
        // ...
    }
}
```

## What I did

I found this article [Symfony Session testing](https://symfony.com/doc/5.4/components/http_foundation/session_testing.html), but it does not explain how to uses these objects.

```php
   // tried to replace this line
   $session = $this->client->getContainer()->get('session');

   // by this and ...
   $session = $this->client->getContainer()->get('request_stack')->getSession();

   // by this and ...
   $session = $this->client->getContainer()->get(RequestStack::class)->getSession();

   // by that
   $session = new Session(new MockFileSessionStorage());
```

## Question
I have no idea how to solve this deprecation nor found any similar example. What would be the correct approach here? Thanks!

**Comment by chyshyya** [comment]
Container get method is deprecated. Use dependency injection:
```php
    public function someControllerAction(RequestStack $requestStack)
    {
        $session = $requestStack->getSession();
        // ...
    }
```

**Reply by DigitalTimK** [reply]
Thanks for your help, but this is not going to work. Dependency Injection (in a function) is only working in a classic controller. Class `WebTestCase` is not extending `AbstractController`

**Reply by chyshyya** [reply]
Pardon. If this is a test then try
```php
$session = new Session(new MockFileSessionStorage());
```
and add `config/packages/tests/framework.yml` with
```yaml
framework:
    test: ~
    # Uncomment this section if you're using sessions
    session:
       storage_id: session.storage.mock_file
```

**Reply by DigitalTimK** [reply]
Thanks again! But unfortunately no success.
I did not need to change `config/packages/tests/framework.yml`, as it was already like this

I replaced the line:
`$session  = $this->client->getContainer()->get('session');` 

with:
`$session = new Session(new MockFileSessionStorage());`

But this is not working, because `$cart`is `null`

Any other idea?

**Reply by chyshyya** [reply]
Try to place
```php
$session = new Session(new MockFileSessionStorage());
```
 to the top of function code. I think that client didn't know about session storage at all when you create one after request.
And set
```php
$client->getCookieJar()->set(new Cookie($session->getName(), $session->getId()));
```
before request

**Reply by DigitalTimK** [reply]
Again many thanks for guiding me through this ... but unfortunately still no success. `$cart`is still `null`.


```php
class BookingControllerTest extends WebTestCase {

    // ...

    public function testToAddItemToCart()
    {
        $session = new Session(new MockFileSessionStorage());

        $this->client  = static::createClient();
        $this->client->getCookieJar()->set(new Cookie($session->getName(), $session->getId()));

        self::ensureKernelShutdown();

        $item = (ProductFactory::random())->object();

        // make REQUEST (which adds stuff into the session)
        $this->client->request('GET', "/cart/add/{$item->getSlug()}");

        // assert RESPONSE
        $this->assertSame(302, $this->client->getResponse()->getStatusCode());
        $this->assertTrue($this->client->getResponse()->isRedirect('/cart/show'));

        // get cart from SESSION
        # $session  = $this->client->getContainer()->get('session');
        $cart     = $session->get('cart');

        // assert content of cart
        // ...
    }
}
```


What am I doing wrong?

**Reply by chyshyya** [reply]
Try this
```php
$session = new Session(new MockFileSessionStorage());
$session->start();
```
It must work

**Reply by DigitalTimK** [reply]
Hey @chyshyya, thanks for your support!!! But it is still not working on my end. So I made a very very basic test-example and also here I did not succeed. 

Maybe you can check this example on your side? Does this work for you?

**CONTROLLER**
```php
<?php

namespace App\Controller;

use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\RedirectResponse;
use Symfony\Component\HttpFoundation\RequestStack;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;

class TestSessionController extends AbstractController
{
    /**
     * @Route("/session_test", name="app_session_test")
     */
    function showSession(RequestStack $requestStack): Response {
        $session = $requestStack->getSession();
        $value_from_session = $session->get("product_id", "no id in session");

        return new Response("Value: {$value_from_session}");
    }

    /**
     * @Route("/session_test/add", name="app_session_test_add")
     */
    function addValueToSession(RequestStack $requestStack): RedirectResponse {
        $session = $requestStack->getSession();
        $session->clear();
        $session->set("product_id", 12345);

        return $this->redirectToRoute('app_session_test');
    }
}
```

**FUNCTIONAL TEST**
```php
<?php

namespace App\Tests\Controller;

use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
use Symfony\Component\HttpFoundation\Session\Session;
use Symfony\Component\HttpFoundation\Session\Storage\MockFileSessionStorage;

class TestSessionControllerTest extends WebTestCase
{
    public function testAddValueToSession(): void
    {
        // prepare SESSION (new way, not working)
        $session = new Session(new MockFileSessionStorage());
        $session->start();

        // start CLIENT
        $client = static::createClient();

        // prepare SESSION (old way - working, but deprecated)
        # $session = $client->getContainer()->get('session');

        // ADD: request & assert RESPONSE
        $client->request('GET', '/session_test/add');
        $this->assertSame(302, $client->getResponse()->getStatusCode());
        $this->assertTrue($client->getResponse()->isRedirect('/session_test'));

        // SHOW: request & assert RESPONSE
        $client->request('GET', '/session_test');
        $this->assertResponseIsSuccessful();
        $this->assertSelectorTextSame('body', 'Value: 12345');

        // assert SESSION
        $product_id = $session->get('product_id');
        $this->assertSame(12345, $product_id);
    }
}
```

**Reply by DigitalTimK** [reply]
Hey @chyshyya, did you had a chance to look at the example? 
Thanks for any support!

**Reply by arderyp** [reply]
Seems like you two were on the cusp of figuring it out. Did you end up figuring it out on your own @DigitalTimK ?

**Reply by DigitalTimK** [reply]
No, unfortunately not successful so far.

**Comment by marien-probesys** [comment, marked as answer]
Hey there, I found a way to access the controller session in my tests. It is heavily based on insights from @chyshyya and from your own code @DigitalTimK, thanks to both of you :pray: 

I have to tell you that I'm using Symfony 6.1 and my solution may not totally fix your case as you're using Symfony 5.4. I hope it will!

There were two missing things in your previous attempts:

1. the cookie domain had to be set (e.g. `localhost`)
2. `new MockFileSessionStorage()` didn't create the session file in the same directory as the one of the controller session (it took me a while to figure it out), which is defined by the `session.save_path` configuration key

I extracted my code in a trait so it can be easily reused in my tests.

```php
<?php

namespace App\Tests;

use Symfony\Bundle\FrameworkBundle\KernelBrowser;
use Symfony\Component\BrowserKit\Cookie;
use Symfony\Component\HttpFoundation\Session\Session;
use Symfony\Component\HttpFoundation\Session\Storage\MockFileSessionStorage;

trait SessionHelper
{
    public function createSession(KernelBrowser $client): Session
    {
        $cookie = $client->getCookieJar()->get('MOCKSESSID');

        // create a new session object
        $container = static::getContainer();
        $session = $container->get('session.factory')->createSession();

        if ($cookie) {
            // get the session id from the session cookie if it exists
            $session->setId($cookie->getValue());
            $session->start();
        } else {
            // or create a new session id and a session cookie
            $session->start();
            $session->save();

            $sessionCookie = new Cookie(
                $session->getName(),
                $session->getId(),
                null,
                null,
                'localhost',
            );
            $client->getCookieJar()->set($sessionCookie);
        }

        return $session;
    }
}
```

Then it can be used this way:

```php
<?php

namespace App\Tests\Controller;

use App\Tests\SessionHelper;
use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;

class SessionControllerTest extends WebTestCase
{
    use SessionHelper;

    public function testUpdateLocaleSavesTheLocaleInTheSession(): void
    {
        $client = static::createClient();
        $session = $this->createSession($client);

        $client->request('POST', '/session/locale', [
            'locale' => 'fr_FR',
        ]);

        $this->assertSame('fr_FR', $session->get('_locale'));
    }
}
```

I hope my solution will help you!

**Reply by DigitalTimK** [reply]
Very cool. Thanks for sharing your code. The trait is perfect! Just by copy/paste ... it worked!

**Reply by marien-probesys** [reply]
I've improved my own solution to be able to retrieve an existing session (started via `loginUser()` for instance) and set a CSRF token. I've added my answer to this other discussion since it asked the exact problem I was facing: https://github.com/symfony/symfony/discussions/46961#discussioncomment-4573371

**Reply by crtl** [reply]
@marien-probesys Your solution was the only one working for me.

**Reply by marien-probesys** [reply]
@crtl Glad it did!

@DigitalTimK I didn't look at this conversation for a while, but it seems that my solution helped many people. May I suggest that you mark my solution as an answer to your question?

**Reply by crtl** [reply]
The first one wasnt working though. Only the second one in the linked comment. Guess its because I also was using authentication.

**Reply by marien-probesys** [reply]
> The first one wasnt working though. Only the second one in the linked comment. Guess its because I also was using authentication.

You did right: the solution I gave here didn't work anymore. I forgot that I changed the solution in the linked comment. Nothing related to authentication :smile: Both solutions should work now!

**Comment by mricherzhagen** [comment]
It should be noted, that `KernelBrowser` emulates a real browser when it comes to Cookies. So when you are dealing with "Session disappears during testing between requests"-issues you should also check, if the Session Cookie was even part of the request.
Depending on the client settings and your request it might not have sent the session cookie. E.g. if you change the `HTTP_HOST` server parameter the above solution won't work, because it sets the cookie only for `localhost`.
So to keep using a session you need to make sure there are no inconsistencies in the hos and protocol you are using. 

If you want to call a service that uses the `Session` you need to simulate a request by inserting a temporary `Request` into the `RequestStack` that has the `Session`, with the ID of your Clients `MOCKSESSID` Cookie, set. And probably `save` the `Session` and `pop` the request afterward.

Using `assertResponseNotHasCookie` everywhere, where you don't expect a cookie, is probably also not a bad idea, so you can be sure, that the session you think is used is still used. We had a bug were the Session-Id kept changing during tests were this would have been valuable.
